http-clients/src/main/java/org/triplea/http/client/web/socket/messages/WebsocketMessageType.java

```diff
@@ -0,0 +1,30 @@
+package org.triplea.http.client.web.socket.messages;
+
+import com.google.common.base.Preconditions;
+import java.util.function.Consumer;
+import java.util.function.Function;
+
+/**
+ * Interface representing a websocket message type that knows how to extract a payload object from a
+ * {@code ServerMessageEnvelope} and also knows how to send that payload to a parameter listener.
+ *
+ * @param <T> Parameterized listener class type. The class is expected to be a data object
+ *     containing {@code Consumer<..>} objects representing per-message-type listeners.
+ */
+public interface WebsocketMessageType<T> {
+  Class<?> getClassType();
+
+  Function<T, Consumer<?>> getListenerMethod();
+
+  @SuppressWarnings("unchecked")
+  default <X> void sendPayloadToListener(
+      final ServerMessageEnvelope serverMessageEnvelope, final T listener) {
+    Preconditions.checkArgument(
+        serverMessageEnvelope.getMessageType().equals(toString()),
+        String.format(
+            "Unexpected message type: %s, wanted message type: %s",
+            serverMessageEnvelope.getMessageType(), toString()));
+    final X payload = (X) serverMessageEnvelope.getPayload(getClassType());
+    ((Consumer<X>) getListenerMethod().apply(listener)).accept(payload);
+  }
```

Comment:
Had to think about our discussion about this, and now that this is a generic interface I really feel like the `<X>` generic type should be part of the interface declaration.
Previously this method was "safe", because the only usages were in the same class, where everything obviously had the same type.
Actually the code could have been changed to guarantee compile-time safety by changing the constructor:
```java
enum Test {
 ABC(String.class, o -> (s -> {}));

 <X> Test(Class<X> classType, Function<T, Consumer<X>> consumer) {
    // Set fields here
 }
}
```

Obviously the underlying code still would've been weird to a certain degree, because in this case were obviously limited by the compiler, but there would be no way of messing up the types at all.

Now that this is an interface there aren't any guarantees that the types returned by `getClassType()` does in any way match the function returned by `getListenerMethod()`, which defeats the purpose of generics and type safety IMO.

However I'm not only here to complain, I think I have a solution that
1. Solves the weird casting issue and
2. Still allows us to use enums

The only limitation is that the generic type "X" can't be checked anywhere except in the implementing class.

Here's the plan (I only tested the important things, so I might have missed a minor detail).

This interface get's a new generic parameter X, and both non-default methods, use that type.
Then you create a new wrapper interface, defining this exact method, implementations are supposed to delegate this call to a `TypeWrapper<T, ?>` (or alternatively a `TypeWrapper<T, ?>` getter), that is implemented by all current `WebsocketMessageType` enums to keep the logic roughly the same.
Then you create a single `TypeWrapper<T, X>` that implements this method here and the changed interface, that is currently being used by all enums.
All enums that currently implement this interface are changed to implement the newly created interface, and only accept a `TypeWrapper` object in their constructor (or create this object based on constructor parameters in a type safe manner), which is then being used to implement the logic here.

I assume this was probably really confusing, so have a look at my test implementation that does this without defining a new interface for enums to implement:
```java
import java.util.function.Consumer;
import java.util.function.Function;

class ServerMessageEnvelope {
 <X> X getPayload(Class<X> clazz) {
   return null;
 }
}

interface WebsocketMessageType<T, X> {
  Class<X> getClassType();

  Function<T, Consumer<X>> getListenerMethod();

  default void sendPayloadToListener(
      final ServerMessageEnvelope serverMessageEnvelope, final T listener) {
    final X payload = serverMessageEnvelope.getPayload(getClassType());
    getListenerMethod().apply(listener).accept(payload);
  }
}

class TypeWrapper<T, X> implements WebsocketMessageType<T, X> {

  private final Class<X> clazz;
  private final Function<T, Consumer<X>> function;

  TypeWrapper(Class<X> clazz, Function<T, Consumer<X>> function) {
    this.clazz = clazz;
    this.function = function;
  }



  @Override
  public Class<X> getClassType() {
    return clazz;
  }

  @Override
  public Function<T, Consumer<X>> getListenerMethod() {
    return function;
  }
}

enum Test {
 ABC(new TypeWrapper<>(Integer.class, a -> (s -> {}))),
 DEF(Double.class, a -> (s -> {}));

 private TypeWrapper<String, ?> wrapper;

 Test(TypeWrapper<String, ?> wrapper) {
   this.wrapper = wrapper;
  }

  <X> Test(Class<X> clazz, Function<String, Consumer<X>> function) {
   this.wrapper = new TypeWrapper<>(clazz, function);
  }

  void sendPayloadToListener(
      final ServerMessageEnvelope serverMessageEnvelope, final String listener) {
    wrapper.sendPayloadToListener(serverMessageEnvelope, listener);
  }
}
```

I'll review the rest of this PR later this day, took my quite some time to check my code idea.

Reply:
Also found this https://openjdk.java.net/jeps/301

Exactly our use-case ^^

Reply:
> Now that this is an interface there aren't any guarantees that the types returned by getClassType() does in any way match the function returned by getListenerMethod(), which defeats the purpose of generics and type safety IMO.

The extraction I believe is equivalent code. It should not have changed any type guarantees. Without a generic typed enum, I suspect we're going to be stuck. We are casting from a JSON String to a type, because the deserialization depends on the serialization, there will be a fundamental 'not type safe', it's unavoidable at some level.

Jeps/301 does look like a match, regrettably it's still just a candidate change though: https://bugs.openjdk.java.net/browse/JDK-8170351


I commend your example @RoiEXLab and trying to fix the type safety issues, though we can only go so far in any one refactoring iteration, I feel like we're getting a bit further from what is 'prudent'. The updates you're describing seem to be more logical/easier once the code has had this first transformation. I think we should take this level of refactor updates to an issue ticket for out-of-band follow-up. The update here is lifting code and not really changing it. We have 2 listeners now, a 3rd is imminent and being worked on, the purpose of this update was to unify the first 2 and pave the way for the third.


### Constructor change

I gave this update a try in the chat message type enum:
```
enum Test {
 ABC(String.class, o -> (s -> {}));

 <X> Test(Class<X> classType, Function<T, Consumer<X>> consumer) {
    // Set fields here
 }
}
```

I was not able to get it to compile. Either the 'T' or the 'Consumer' type conflicted.


### TypeWrapper
The TypeWrapper could have possibility. Though I'm a bit confused on:

```
 private TypeWrapper<String, ?> wrapper;
```

Why is that as String?


```
 ABC(new TypeWrapper<>(Integer.class, a -> (s -> {}))),
 DEF(Double.class, a -> (s -> {}));
```

Why does Double not need a type wrapper but integer does?


### Follow up

It seems we're blocked, the above comments are awkward as I can't comment on a diff. I think it's evidence we're looking at further refactorings. Would you mind explaining excactly what you think should be changed that is within the intent of this update @RoiEXLab? Please also clarify if you think that extracting common code to a base-class actually does change type-safety, I'm curious how that would work as the classes shoudl be equivalent.

Last, I'm concerned about the test/brittleness effort. I spent about 60 minutes last night debugging chat after this refactor went too far and I had to roll back a number of updates to get it working again. As mentioned, because we deserialize from JSON, the code is not going to be fully type-safe. So I don't think we can fully solve it, it's fundamental to the JSON deserialization.

Reply:
I updated my original comment to give some hints, but it's really hard explaining it.
That's why I created a PR onto your branch to make it more clear: https://github.com/DanVanAtta/triplea/pull/7

Reply:
Thanks @RoiEXLab , though perhaps you can appreciate that the update you are suggesting is significant.

The trade-off made here was roughly as follows:
- fixes duplication
- unifies the pattern for listeners
- removes testing burden and simplifies
- allows the next listener to drop in with nicer tests, no throw-away work for unnecessary tests, and consistent to the above pattern
- type non-safety added (IMO it's annoying, but not egregious)